plugin/mlsdbprovider.cpp

```diff
@@ -134,51 +134,74 @@ MlsdbProvider::~MlsdbProvider()
-                qCDebug(lcGeoclueMlsdb) << "geoclue-mlsdb data file" << fname << "version unknown:" << version;
-                continue; // ignore this file
+// TODO: Search alternative locations for files with mlsdb data
+bool MlsdbProvider::searchForCellIdLocation(quint64 uniqueCellId, MlsdbCoords *coords) {
+    if (uniqueCellId == 0) {
+        return false;
+    }
+    QString df = "/usr/share/geoclue-provider-mlsdb/data/";
+    QString lf = "/usr/share/geoclue-provider-mlsdb/data/";
+    quint16 mcc = getCellMcc(uniqueCellId);
+    df.append(QString::number(mcc));
+    lf.append(QString::number(mcc));
+    df.append(".dat");
+    df.append(".loc");
+    quint64 network = 0;
+    quint64 uid = uniqueCellId & 0xFFFFFFFFFFFFFF;
+    FILE *fp;
+    size_t fs, pos, half;
+    size_t old_pos = 1;
+    fp = fopen(df.toStdString().c_str(), "r");
+    if (fp == NULL) {
+        fprintf(stderr, "WARNING: Unable to open data file for mcc %d\n", mcc);
+        return false;
+    }
+    fseek(fp, 0, SEEK_END);
+    fs = ftell(fp);
+    if (fs % DATA_SIZE != 0) {
+        fprintf(stderr, "ERROR: File size is not a multiple of data size. Corrupt file?\n");
+        return false;
+    }
+    pos = fs / 2;
+    if (pos % DATA_SIZE != 0) {
+        pos -= DATA_SIZE / 2;
+    }
+    half = pos;
+    while (uid != network) {
+        if (old_pos == pos) {
+            fprintf(stderr, "WARNING: Could not find exact record. Target is %llu, closest match is %llu\n", uid, network);
+            return false;
+        }
+        old_pos = pos;
+        fseek(fp, pos, SEEK_SET);
+        fread(&network, DATA_SIZE, 1, fp);
+        half /= 2;
+        if (half % DATA_SIZE != 0) {
+            half -= DATA_SIZE / 2;
+            if (half == 0) {
+                half = DATA_SIZE;
             }
-
-            QMap<MlsdbUniqueCellId, MlsdbCoords> perLcCellIdToLocations;
-            in >> perLcCellIdToLocations;
-            if (perLcCellIdToLocations.isEmpty()) {
-                qCDebug(lcGeoclueMlsdb) << "geoclue-mlsdb data file" << fname << "contained no cell locations!";
-            } else {
-                if (perLcCellIdToLocations.contains(uniqueCellId)) {
-                    *coords = perLcCellIdToLocations.value(uniqueCellId);
-                    qCDebug(lcGeoclueMlsdb) << "geoclue-mlsdb data file" << fname << "contains the location of composed cell id:" << uniqueCellId.toString() << "->" << coords->lat << "," << coords->lon;
-                    return true; // found!
-                } else {
-                    qCDebug(lcGeoclueMlsdb) << "geoclue-mlsdb data file" << fname << "contains" << perLcCellIdToLocations.size() << "cell locations, but not for:" << uniqueCellId.toString();
-                }
+        }
+        if (uid > network) {
+            pos += half;
+        }
+        else if (uid < network) {
+            pos -= half;
+        }
+        else {
+            fclose(fp);
+            fp = fopen(lf.toStdString().c_str(), "r");
```

Comment:
Two files per data set is already better, but pondering would it be still possible to merge the two. E.g. some simple header for content count or offset of the other data. For bonus that could provide a bit more robustness or the handled files (= file is really of correct content and possibly version).